subscription/src/main/java/org/killbill/billing/subscription/api/svcs/DefaultSubscriptionInternalApi.java

```diff
@@ -267,24 +269,25 @@ public SubscriptionBase createSubscription(final UUID bundleId, final PlanPhaseS
 
                 final SubscriptionBaseBundle bundle = createBundleForAccount(accountId, entitlementWithAddOnsSpecifier.getExternalKey(), context);
 
-                final List<SubscriptionSpecifier> subscriptions = verifyAndBuildSubscriptionSpecifiers(bundle.getId(),
-                                                                                                       entitlementWithAddOnsSpecifier.getEntitlementSpecifier(),
-                                                                                                       entitlementWithAddOnsSpecifier.isMigrated(),
-                                                                                                       context,
-                                                                                                       now,
-                                                                                                       effectiveDate,
-                                                                                                       catalog,
-                                                                                                       callContext);
-
-                SubscriptionAndAddOnsSpecifier subscriptionAndAddOnsSpecifier = new SubscriptionAndAddOnsSpecifier();
-                subscriptionAndAddOnsSpecifier.setSubscriptionSpecifiers(subscriptions);
+                SubscriptionAndAddOnsSpecifier subscriptionAndAddOnsSpecifier = new SubscriptionAndAddOnsSpecifier(
+                        bundle.getId(),
+                        effectiveDate,
+                        verifyAndBuildSubscriptionSpecifiers(bundle.getId(),
+                                                             entitlementWithAddOnsSpecifier.getEntitlementSpecifier(),
+                                                             entitlementWithAddOnsSpecifier.isMigrated(),
+                                                             context,
+                                                             now,
+                                                             effectiveDate,
+                                                             catalog,
+                                                             callContext)
+                );
                 subscriptionAndAddOns.add(subscriptionAndAddOnsSpecifier);
             }
 
-            final List<DefaultSubscriptionBase> result = apiService.createPlansWithAddOns(accountId, subscriptionAndAddOns, callContext);
-            return ImmutableList.copyOf(Iterables.transform(result, new Function<DefaultSubscriptionBase, SubscriptionBase>() {
+            final List<DefaultSubscriptionBaseWithAddOns> result = apiService.createPlansWithAddOns(accountId, subscriptionAndAddOns, callContext);
+            return ImmutableList.copyOf(Iterables.transform(result, new Function<DefaultSubscriptionBaseWithAddOns, SubscriptionBaseWithAddOns>() {
```

Comment:
@pierre 
>
My original comment was about rewriting the code to use a simple for loop instead of ImmutableList.copyOf(Iterables.transform in a few places (easier to read, no need for the copyOf as the items can be added in a list right away, etc.).
>

I've just pushed the code using the for loop.

>
But now, I am confused why the conversion is even needed since DefaultSubscriptionBaseWithAddOns should implement SubscriptionBaseWithAddOns already.
>

I too am confused about that. But if I don't convert it I'm getting the "Incompatible types" error in the compiler:
Required `List<SubscriptionBaseWithAddOns>`, found `List<DefaultSubscriptionBaseWithAddOns>` 🤔